packages/snjs/lib/services/history/history_manager.ts

```diff
@@ -33,189 +54,259 @@ const PERSIST_TIMEOUT = 2000;
-    this.sessionHistory = SessionHistoryMap.FromJson(
-      this.storageService!.getValue(StorageKey.SessionHistoryRevisions)
+    this.history = await this.getPersistedHistory();
+    this.autoOptimize = await this.storageService.getValue(
+      StorageKey.SessionHistoryOptimize,
+      undefined,
+      true
     );
-    const autoOptimize = await this.storageService!.getValue(
-      StorageKey.SessionHistoryOptimize
+    this.addChangeObserver();
+  }
+
+  private async getPersistedHistory(): Promise<
+    Record<UuidString, HistoryEntry[]>
+  > {
+    const historyMap: Record<UuidString, HistoryEntry[]> = {};
+    const rawHistory: PersistableHistory = await this.storageService.getValue(
+      StorageKey.SessionHistoryRevisions
     );
-    if (isNullOrUndefined(autoOptimize)) {
-      /** Default to true */
-      this.autoOptimize = true;
-    } else {
-      this.autoOptimize = autoOptimize;
+    if (!rawHistory) {
+      return historyMap;
     }
-    this.addChangeObserver();
+    for (const [uuid, rawHistoryArray] of Object.entries(rawHistory)) {
+      const entries: HistoryEntry[] = [];
+      for (const [index, rawEntry] of rawHistoryArray.entries()) {
+        const payload = CreateSourcedPayloadFromObject(
+          rawEntry.payload,
+          PayloadSource.SessionHistory
+        ) as SurePayload;
+        const previousEntry = entries[index + 1];
+        const entry = CreateHistoryEntryForPayload(payload, previousEntry);
+        entries.unshift(entry);
+      }
+      historyMap[uuid] = entries;
+    }
+    return historyMap;
   }
 
-  addChangeObserver() {
-    this.removeChangeObserver = this.itemManager!.addObserver(
-      this.contentTypes,
-      (changed, inserted, discarded, _ignored, source) => {
-        const items = concatArrays(changed, inserted, discarded) as SNItem[];
-        if (source === PayloadSource.LocalChanged) {
-          return;
-        }
-        for (const item of items) {
-          try {
-            if (!item.deleted && !item.errorDecrypting) {
-              this.addHistoryEntryForItem(item);
-            }
-          } catch (e) {
-            console.error('Unable to add item history entry:', e);
-          }
+  private addChangeObserver() {
+    this.removeChangeObserver = this.itemManager.addObserver(
+      this.historyTypes,
+      (changed) => {
+        if (changed.length > 0) {
+          this.recordNewHistoryForItems(changed);
         }
       }
     );
   }
 
+  private recordNewHistoryForItems(items: SNItem[]) {
+    let needsPersist = false;
+    for (const item of items) {
+      if (!this.historyTypes.includes(item.content_type)) {
+        continue;
+      }
+      const payload = item.payload;
+      if (
+        item.deleted ||
+        payload.format !== PayloadFormat.DecryptedBareObject
+      ) {
+        continue;
+      }
+      const itemHistory = this.history[item.uuid] || [];
+      const latestEntry = itemHistory.length > 0 ? itemHistory[0] : undefined;
```

Comment:
So this one was really strange.

```
const latestEntry = itemHistory.length > 0 ? itemHistory[0] : undefined;
```

Above `latestEntry` has type `HistoryEntry | undefined`, which is what I want.

```
const latestEntry = itemHistory[0];
```

And here `latestEntry` has type `HistoryEntry`, which is not what I want.

Reply:
Oh wow yeah, that is really strange. Maybe it's a TypeScript default because it would be a pain in other instances?

I guess you could always do this:
```ts
 const latestEntry: HistoryEntry | undefined = itemHistory[0];
 ```
 
 But at that point it's about as busy as your original line so 🤷‍♀️